### PR TITLE
Print backtraces of threads and threadpools - MOD-5269 

### PR DESCRIPTION
**Pull request by meiravgri**
This PR introduces an API to get backtraces of threads that exist in the process.

The first step that the process has to do in order to support this ability is to sign in self to the signal handler that pauses the threads using `register_process_to_pause_handler()`

We have two types of implementations:
1. Get the backtraces from **all threads that exist** in the process,  which does the following:
* For each thread id obtained from from `\proc\<pid>\tasks` directory, assuming that the Redis process is the file owner and hence has reading permission for it, send a signal to pause it, excluding the caller and including the main thread (if it's not the caller). The threads that didn't block or ignore the signal immediately call the signal handler and wait until the buffer to write the backtrace to is ready.
* Go over each `\proc\<pid>\tasks\<tid>\status` file to get the threads' signals mask, to check how many threads actually got the signal.
* Initialize a buffer for all the threads' backtraces and turn on a flag that allows the threads to start writing.
* Wair for the threads that were paused to write to the buffer and write the caller backtrace to the buffer as well.
* Print the backtraces to the info context (in case of crash) or to a reply (in case it was called by ft.debug)
* resume the threads
** NOTE: We are using system calls and the `\proc` directory, which are supported only in linux!
** NOTE2: The threads' order in the output is random.

2. Get backtrace of a specific threadpool
* Pause the requested thpool
* Initialize the buffer
* let the thread continue to the writing phase
* wait until all threads done writing
* print the backtraces
* resume the threadpool

** For this implementations we use pthread POSIX API, so it is limited to be used in linux.
As we are using global variables and structures, for both use cases we need to first check that there is no dump process already running with `ThpoolDump_test_and_start()`. To mark we are done with the dump process call `ThpoolDump_finish()`

On crash, if we are in linux platform,  we call `ThpoolDump_all_to_info` and the backtraces of the threads that got the signal will be added to the crash log.

To get the backtraces at runtime we introduce a new `FT.DEBUG` command:
`FT.DEBUG DUMP_THREADPOOL_BACKTRACE [thpool_name]`
there thpool_name can be :
ALL - calls the Linux-supported implementation.
or one of the following: GC, SEARCH_0, CLEANSPEC, WORKERS
Will dump the backtrace of the requested thpool. Supported on all platforms.

Additional changes in this PR:
1. Pause and resume functions in thpool API are actually general. If it will be found useful for whatever use case it can be exposed by wrapping it in a specific RediSearch thpool API.
2. The thpools and threads are given a name
The thpool name is stored in the thpool struct and limited to 10 bytes (if it is longer will be truncated). If the platform supports it, the thread name is stored in the system. Threads' names are restricted to 16 bytes and will have the following format <thpool_name>-<thread_id>. The thread name is used as a title to its backtrace.
4. Logging function in the thpool API was expanded to get a variadic number of arguments to allow formated log.

### CODE REVIEW

File: deps/thpool/thpool.c [hunk 1]
```diff
@@ -88,20 +97,37 @@ typedef struct redisearch_thpool_t {
...
+
+static volatile bool g_pause_all = 0;
+
+static volatile bool g_resume_all = 1;
+
 /* ========================== PROTOTYPES ============================ */
 
-static int thread_init(redisearch_thpool_t* thpool_p, struct thread** thread_p, int id, LogFunc log);
+// Cleanup allocations related to the threadpool
+static void redisearch_thpool_cleanup(redisearch_thpool_t* thpool_p);
```
Comment (MeirShpilraien) [inline deps/thpool/thpool.c:122]: Why the `redisearch_` prefix?

File: deps/thpool/thpool.c [hunk 2]
```diff
@@ -121,50 +147,120 @@ static void bsem_post(struct bsem* bsem_p);
...
-/* ========================== THREADPOOL ============================ */
+static bool pause_all() {
+  return g_pause_all;
+}
+static bool resume_all() {
+  return g_resume_all;
+}
+/* ========================== GENERAL ============================ */
+/* Register the process to a signal handler */
+void register_process_to_pause_handler(LogFunc log_cb) {
```
Comment (MeirShpilraien) [inline deps/thpool/thpool.c:158]: I think this should be located somewhere on RediSearch code no? Not on thpool.c?

File: deps/thpool/thpool.c [hunk 3]
```diff
@@ -121,50 +147,120 @@ static void bsem_post(struct bsem* bsem_p);
...
+  /* Register signal handler */
+  struct sigaction act;
+  struct sigaction oldact;
+  sigemptyset(&act.sa_mask);
+  act.sa_flags = 0;
+  act.sa_handler = thread_hold;
+  if (sigaction(SIGUSR2, &act, &oldact) == -1) {
```
Comment (MeirShpilraien) [inline deps/thpool/thpool.c:165]: We need to make sure Redis is not using SIGUSR2.

File: deps/thpool/thpool.c [hunk 4]
```diff
@@ -121,50 +147,120 @@ static void bsem_post(struct bsem* bsem_p);
...
+  if (oldact.sa_handler != SIG_DFL) {
+    log_cb("register_process_to_pause_handler(): changing SIGUSR2 handler");
+  }
+}
 
-/* Create thread pool */
-struct redisearch_thpool_t* redisearch_thpool_create(size_t num_threads) {
-  threads_on_hold = 0;
+#if defined(__linux__)
+size_t pause_all_process_threads() {
```
Comment (MeirShpilraien) [inline deps/thpool/thpool.c:176]: Same comment as above, I do not think this should be on thpool.c

File: deps/thpool/thpool.c [hunk 5]
```diff
@@ -175,16 +271,15 @@ struct redisearch_thpool_t* redisearch_thpool_create(size_t num_threads) {
...
-void redisearch_thpool_init(struct redisearch_thpool_t* thpool_p, LogFunc log) {
-  assert(thpool_p->keepalive == 0);
-  thpool_p->keepalive = 1;
-  thpool_p->terminate_when_empty = 0;
-
+/* initialize thread pool */
+void redisearch_thpool_init(struct redisearch_thpool_t* thpool_p) {
+  assert(!(thpool_p->flags & RS_THPOOL_F_KEEP_ALIVE));
+  redisearch_thpool_TURNON_flag(thpool_p, RS_THPOOL_F_KEEP_ALIVE);
+  redisearch_thpool_TURNOFF_flag(thpool_p, RS_THPOOL_F_TERMINATE_WHEN_EMPTY);
```
Comment (MeirShpilraien) [inline deps/thpool/thpool.c:278]: Why we need the `redisearch_` prefix? I look at thpool.c as a general thread pool implementation so we should avoid module specific names.

File: deps/thpool/thpool.c [hunk 6]
```diff
@@ -341,33 +436,128 @@ void redisearch_thpool_destroy(redisearch_thpool_t* thpool_p) {
...
+
+  while (thpool_p->flags & RS_THPOOL_F_RESUME) {
+    sleep(1);
+    LOG_FMT_IF_EXISTS("redisearch_thpool_pause(): waiting for %s threadpool to finish resuming process.",
+        thpool_p->name);
+  }
   size_t n;
-  for (n = 0; n < thpool_p->num_threads_alive; n++) {
-    pthread_kill(thpool_p->threads[n]->pthread, SIGUSR2);
+	pthread_t caller = pthread_self();
```
Comment (MeirShpilraien) [inline deps/thpool/thpool.c:454]: Indentation issue?

File: deps/thpool/thpool.c [hunk 7]
```diff
@@ -341,33 +436,128 @@ void redisearch_thpool_destroy(redisearch_thpool_t* thpool_p) {
...
+
+  // set hold flag
+  redisearch_thpool_TURNON_flag(thpool_p, RS_THPOOL_F_PAUSE);
+
+  // zero the number of paused threads in the current threadpool.
+  // This number will be increased atomically by each thread to be used
+  // as an unique index in the output array.
+  g_threads_paused_cnt = 0;
+
+	for(n = 0; n < threadpool_size; n++) {
```
Comment (MeirShpilraien) [inline deps/thpool/thpool.c:471]: Indentation?

File: deps/thpool/thpool.c (hunk 5 above)
Comment (rafie) [reply deps/thpool/thpool.c:278]: @MeirShpilraien This is due to use of RediSearch within RedisGraph, to avoid namespace clashes.

**Pull request closed without merging**